### ISSUE TITLE
Build a decent test case reducer

### ISSUE THREAD

**Issue by agroce** [trailofbits/deepstate#109]
https://github.com/agroce/testfs/createreset.test and .out
https://github.com/agroce/testfs/rmdirreset.test and .out

These are the shortest test cases for these FATALs found by libFuzzer in an overnight fuzzing run.  I am fairly sure they are not close to shortest-possible (for example, are those failed operations really needed?).  We should build something like the reducer in Hypothesis https://github.com/HypothesisWorks/hypothesis, where reduction works on the bytestream itself, but intelligently.  Alternatively, there could be a custom libFuzzer mutator that implements smart test case reduction, perhaps aware of OneOf boundaries.

**Comment by pgoodman** [comment]
so I mean, there is a second logical stream which we ignore, and that is the sequence of api calls into deepstate which we make, and which provide some structure to the test cases. Some components are irreducible, though we have no way of telling that to downstream users. We might be able to provide hints. Some ideas:

So, `OneOf` is basically requesting an integer, which is an index into an array of actions to perform. Thus, the symbolic integers returned have a semantic: they are choices, and they have ranges, and a reducer that understands this semantic can apply a reduction strategy that is about re-ordering and/or choosing a subset of the choices. So how do we get the reducer this info?
 
There's a few ways, but basically, the bytestream is a tree, and when you call `OneOf`, you want to tell the system that you're making a choice, that some integer represents the choice, and then every following thing up until the chosen callback invoked by `OneOf` returns is a result of this choice. 

This might also mean that we want a more structured pattern than just `OneOf`: we want a loop of `OneOf`s, and therefore provide the reducer the ability to recognize a sequence of related choices.

**Comment by agroce** [comment]
So in some sense this is what Hypothesis does, and is what I'm hacking up a very shallow version of.

What does the second mean?  Right now, the structure of bytes in a single OneOf call is clear enough.  And it can presumably figure out nesting, so what is the new thing made visible by the looping OneOf?

**Comment by agroce** [comment]
As a first stab at seeing how this does, I'm making a branch that makes byte reads and OneOf boundaries explicit via the dumb channel of I/O, but this can drive some Python code that reduces tests.  For libFuzzer, we want a mode where there's an ability to run a test case with a structure passed in that records the structure of that run, so the mutator can use that to pick a plausible reduction (and at the same time can see the point at which an assumption is violated).

**Comment by pgoodman** [comment]
Lets say you have this...

```c++
for (int i = 0; i < 5; ++i) {
  OneOf(
    [=]() { symbolic_int j; ASSUME(j > 0); foo(j); },
    [=]() { symbolic_int j; ASSUME(j == 0); foo(j); },
    [=]() { symbolic_int j; ASSUME(j < 0); foo(j); } 
  );
}
```

So what you would get is `5 * (sizeof(int) + sizeof(int))` symbolic bytes. What you really have though is a tree structure...

```
(choices
  (choice_int
    (symbolic_int))
  (choice_int
    (symbolic_int))
  (choice_int
    (symbolic_int))
  (choice_int
    (symbolic_int))
  (choice_int
    (symbolic_int)))
```

**Comment by pgoodman** [comment]
Right now we have no way of conveying that the various choices are related under `choices`. There are some ways we could go about this....

We could add a ranged loop construct, which would have the benefit of letting us inject in pumps without actually making them explicit, and then a new fuzzing approach could be to fuzz on a pumping strategy, and then if you see oneof choices inside a loop, you can assume they are related. This is kind of cool I think...

**Comment by pgoodman** [comment]
Well is pumping meaningful for fuzzing.. I dunno.

**Comment by agroce** [comment]
What does Pump do in the fuzzer, anyway?  It's a NOP there, right?

**Comment by agroce** [comment]
I see, you could make it clear that you will always have 5 of these choices.

**Comment by agroce** [comment]
Ok!  https://github.com/agroce/testfs/blob/master/reducer.py shows that you can write a really simple dumb reducer and get something out of it.  Also makes it clear the reset vulnerability in testfs is not completely trivial -- you need to create a dir, then remove a dir, then do a create with a well-timed reset to cause abort on mount.

**Comment by agroce** [comment]
Ok, current reducer can do two things, which prove quite useful:

- Get rid of a OneOf, including in a nested structure, by just removing all bytes read during that OneOf from the test.

- Reduce any byte to some lower value

**Comment by agroce** [comment]
Tidied up, this might be worth including as deepstate-reduce in the release, actually.  It's simple but effective, for fuzzer results anyway (I am guessing symbolic execution is going to produce closer-to-minimal tests to start with)

**Comment by agroce** [comment]
The naive reducer isn't bad.  Compare https://github.com/agroce/testfs/blob/master/createreset.out and https://github.com/agroce/testfs/blob/master/mincreatereset.out.  Running the reducer looks like this:

```shell
python reducer.py ./Tests createreset.test mincreatereset.test "FATAL: /root/testfs/super.c(252): Assertion (testfs_inode_get_type(in) == I_FILE) || (testfs_inode_get_type(in) == I_DIR) failed in function int testfs_checkfs(struct super_block *, struct bitmap *, struct bitmap *, int)"
ORIGINAL TEST HAS 119 BYTES
LAST BYTE READ IS 123
ONEOF REMOVAL SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
BYTE REDUCTION SUCCEEDED: 103
ONEOF REMOVAL SUCCEEDED: 91
BYTE REDUCTION SUCCEEDED: 91
ONEOF REMOVAL SUCCEEDED: 87
ONEOF REMOVAL SUCCEEDED: 75
ONEOF REMOVAL SUCCEEDED: 55
ONEOF REMOVAL SUCCEEDED: 51
NO REDUCTIONS FOUND
```

**Comment by agroce** [comment]
Notice also fuzzer-created tests (like this one) can rely on our implict zeros beyond the end of the data.  It might be nice if the reducer also padded out the test with zeros where zeros are getting read, because I can see that causing troubles...

**Comment by agroce** [comment]
Pull request #111 does the external tool, non-libFuzzer, version of this.

**Comment by agroce** [comment]
Fixed with merge of #111 -- future work is to go from decent reducer --> really good reducer :)
